src/register/pmpcfgx.rs

```diff
@@ -1,7 +1,124 @@
 /// Physical memory protection configuration
+
 pub mod pmpcfg0 {
-    read_csr_as_usize!(0x3A0, __read_pmpcfg0);
-    write_csr_as_usize!(0x3A0, __write_pmpcfg0);
+    use bit_field::BitField;
+
+    #[derive(Clone, Copy, Debug)]
+    pub enum Permission {
+        NONE = 0,
+        R = 1,
+        W = 2,
+        RW = 3,
+        X = 4,
+        RX = 5,
+        WX = 6,
+        RWX = 7,
+    }
+
+    #[derive(Clone, Copy, Debug)]
+    pub enum Range {
+        OFF = 0,
+        TOR = 1,
+        NA4 = 2,
+        NAPOT = 3,
+    }
+
+    #[derive(Clone, Copy, Debug)]
+    pub struct Pmpconfig {
+        pub permission: Permission,
+        pub range_type: Range,
+        pub locked: bool,
+    }
+
+    #[derive(Clone, Copy, Debug)]
+    pub struct Pmpcfg0 {
+        bits: usize,
+    }
+
+    impl Pmpcfg0 {
+        ///Returns the pmp byte associated with the index
+        #[inline]
+        pub fn pmp_byte(&self, index: usize) -> usize {
+            #[cfg(riscv32)]
+            assert!(index < 4);
+
+            #[cfg(riscv64)]
+            assert!(index < 8);
+
+            self.bits.get_bits(8 * index..8 * (index + 1))
+        }
+
+        #[inline]
+        fn range(&self, byte: usize) -> Range {
+            match byte.get_bits(4..6) {
+                0 => Range::OFF,
+                1 => Range::TOR,
+                2 => Range::NA4,
+                3 => Range::NAPOT,
+                _ => unreachable!(),
+            }
+        }
+
+        #[inline]
+        fn permission(&self, byte: usize) -> Option<Permission> {
+            match byte.get_bits(0..3) {
+                0 => Some(Permission::NONE),
+                1 => Some(Permission::R),
+                2 => Some(Permission::W),
+                3 => Some(Permission::RW),
+                4 => Some(Permission::X),
+                5 => Some(Permission::RX),
+                6 => Some(Permission::WX),
+                7 => Some(Permission::RWX),
+                _ => None,
+            }
```

Comment:
You should return `None` if `A` is set to `OFF`, but there is no check for the value of the `A` bit.